File: ocaml/AC20.ml
type kind =
  FlipFlop | Conjunction | Broadcaster

type module_description = {
    kind : kind;
    destinations : Common.Atom.t list;
  }

let split_kind_and_name kind_and_name =
  if kind_and_name = "broadcaster" then
    Broadcaster, kind_and_name
  else
    begin
      let name =
        String.sub kind_and_name 1 (String.length kind_and_name - 1) in
      let kind =
        match kind_and_name.[0] with
        | '%' -> FlipFlop
        | '&' -> Conjunction
        | _ -> failwith "split_kind_and_name" in
      kind, name
    end

let get_destination s =
  Common.Atom.get (String.trim s)

let module_description_of_string s =
  let kind_and_name, destinations =
    Scanf.sscanf s "%s -> %s@\n" (fun kind_and_name destinations ->
        kind_and_name,
        List.map get_destination (String.split_on_char ',' destinations)) in
  let kind, name = split_kind_and_name kind_and_name in
  Common.Atom.get name, { kind; destinations }

type pulse = Low | High

module State = struct
  type t = {
      flip_flop : Common.AtomSet.t;
      conjunction : Common.AtomSet.t Common.AtomMap.t;
    }

  let initial input_map = {
      flip_flop = Common.AtomSet.empty;
      conjunction = input_map;
    }
end

type message = {
    source : Common.Atom.t;
    target : Common.Atom.t;
    pulse : pulse;
  }

let flip_flop send { target; pulse; _ } (state : State.t) =
  match pulse with
  | High -> state
  | Low ->
     if Common.AtomSet.mem target state.flip_flop then
       begin
         send Low;
         { state with flip_flop = Common.AtomSet.remove target state.flip_flop }
       end
     else
       begin
         send High;
         { state with flip_flop = Common.AtomSet.add target state.flip_flop }
       end

let conjunction send { source; target; pulse } (state : State.t) =
  let low_set = Common.AtomMap.find target state.conjunction in
  let low_set', changed =
    match pulse, Common.AtomSet.mem source low_set with
    | Low, false -> Common.AtomSet.remove source low_set, true
    | High, true -> Common.AtomSet.add source low_set, true
    | Low, true
    | High, false -> low_set, false in
  let state =
    if changed then
      { state with conjunction = Common.AtomMap.add target low_set' state.conjunction }
    else
      state in
  if Common.AtomSet.is_empty low_set' then
    send Low
  else
    send High;
  state

let button = Common.Atom.get "button"

let broadcaster = Common.Atom.get "broadcaster"

let push_button observe configuration state =
  let pulse_queue = Queue.create () in
  let push ~source ~target pulse =
    observe ~source ~target pulse;
    Queue.push { source; target; pulse } pulse_queue in
  let rec loop state =
    match Queue.pop pulse_queue with
    | exception Queue.Empty -> state
    | message ->
       let state =
         match Common.AtomMap.find message.target configuration with
         | exception Not_found -> state
         | description ->
           let send pulse =
             let source = message.target in
             let send_to target =
               push ~source ~target pulse in
             List.iter send_to description.destinations in
           match description.kind with
           | FlipFlop ->
              flip_flop send message state
           | Conjunction ->
              conjunction send message state
           | Broadcaster ->
              send message.pulse;
              state in
       loop state in
  push ~source:button ~target:broadcaster Low;
  loop state

let add_input source input_map target =
  let inputs =
    try
      Common.AtomMap.find target input_map
    with Not_found ->
      Common.AtomSet.empty in
  Common.AtomMap.add target (Common.AtomSet.add source inputs) input_map

let add_destinations source description input_map =
  List.fold_left (add_input source) input_map description.destinations

let get_input_map configuration =
  Common.AtomMap.fold add_destinations configuration Common.AtomMap.empty

let rec iterate stop_predicate observe configuration input_map state =
  if stop_predicate () then
    state
  else
    begin
      let state' = push_button observe configuration state in
      iterate stop_predicate observe configuration input_map state'
    end

let part1 configuration input_map =
  let counter_low = ref 0 in
  let counter_high = ref 0 in
  let observe ~source:_ ~target:_ pulse =
    match pulse with
    | Low -> incr counter_low
    | High -> incr counter_high in
  let counter = ref 0 in
  let stop_predicate () =
    if !counter < 1000 then
      begin
        incr counter;
        false
      end
    else
      true in
  let _ =
    iterate stop_predicate observe configuration input_map
      (State.initial input_map) in
  !counter_low * !counter_high

let find_first_high_pulses configuration input_map sources =
  let indices = ref Common.AtomMap.empty in
  let counter = ref 0 in
  let stop_predicate () =
    incr counter;
    Common.AtomSet.for_all (fun source -> Common.AtomMap.mem source !indices) sources in
  let observe ~source ~target:_ pulse =
    if
      pulse = High && Common.AtomSet.mem source sources &&
        not (Common.AtomMap.mem source !indices)
    then
      indices := Common.AtomMap.add source !counter !indices in
  let _ =
    iterate stop_predicate observe configuration input_map
      (State.initial input_map) in
  !indices

let part2 configuration input_map =
  let rx = Common.Atom.get "rx" in
  let rx_sources = Common.AtomMap.find rx input_map in
  assert (Common.AtomSet.cardinal rx_sources = 1);
  let kh = Common.AtomSet.choose rx_sources in
  let kh_configuration = Common.AtomMap.find kh configuration in
  assert (kh_configuration.kind = Conjunction);
  let kh_sources = Common.AtomMap.find kh input_map in
  let indices = find_first_high_pulses configuration input_map kh_sources in
  List.fold_left Common.lcm 1 (List.map snd (Common.AtomMap.bindings indices))

let () =
  let configuration =
    Common.AtomMap.of_seq
      (Seq.map module_description_of_string (Common.input_lines ())) in
  let input_map = get_input_map configuration in
  Printf.eprintf "Part 1 : %d\n" (part1 configuration input_map);
  Printf.eprintf "Part 2 : %d\n" (part2 configuration input_map)

File: rust/src/bin/ac20.rs
use advent_of_code::{Result, NameTable, Name, NameSet, NameMap, lcm};

#[derive(Copy, Clone, PartialEq, Eq)]
enum Pulse { Low, High }

trait Module {
    fn add_input(&mut self, input: Name);

    fn get_inputs(&self) -> Option<&NameSet>;

    fn run(&self) -> Box<dyn RunningModule>;
}

trait RunningModule {
    fn receive(
        &mut self, message: &Message, send: &mut dyn FnMut(Pulse)
    );
}

struct FlipFlop {}

struct RunningFlipFlop {
    state: bool,
}

struct Conjunction {
    inputs: NameSet,
}

impl Conjunction {
    fn new() -> Self {
        Self { inputs: NameSet::new() }
    }
}

struct RunningConjunction {
    low_set: NameSet,
}

struct Broadcaster {}

impl Module for FlipFlop {
    fn add_input(&mut self, _input: Name) {}

    fn get_inputs(&self) -> Option<&NameSet> {
        None
    }

    fn run(&self) -> Box<dyn RunningModule> {
        Box::new(RunningFlipFlop { state: false })
    }
}

impl RunningModule for RunningFlipFlop {
    fn receive(&mut self, message: &Message, send: &mut dyn FnMut(Pulse)) {
        if message.pulse == Pulse::Low {
            send(if self.state { Pulse::Low } else { Pulse::High });
            self.state = !self.state;
        }
    }
}

impl Module for Conjunction {
    fn add_input(&mut self, input: Name) {
        self.inputs.insert(input);
    }

    fn get_inputs(&self) -> Option<&NameSet> {
        Some(&self.inputs)
    }

    fn run(&self) -> Box<dyn RunningModule> {
        Box::new(RunningConjunction { low_set: self.inputs.clone() })
    }
}

impl RunningModule for RunningConjunction {
    fn receive(&mut self, message: &Message, send: &mut dyn FnMut(Pulse)) {
        match (message.pulse, self.low_set.contains(&message.input)) {
            (Pulse::Low, false) => { self.low_set.insert(message.input); }
            (Pulse::High, true) => { self.low_set.remove(&message.input); }
            (Pulse::Low, true) | (Pulse::High, false) => (),
        };
        send(if self.low_set.is_empty() { Pulse::Low } else { Pulse :: High })
    }
}

impl Module for Broadcaster {
    fn add_input(&mut self, _input: Name) {}

    fn get_inputs(&self) -> Option<&NameSet> {
        None
    }

    fn run(&self) -> Box<dyn RunningModule> {
        Box::new(Broadcaster {})
    }
}

impl RunningModule for Broadcaster {
    fn receive(&mut self, message: &Message, send: &mut dyn FnMut(Pulse)) {
        send(message.pulse);
    }
}

struct ModuleDescription<T> {
    module: T,
    destinations: Vec<Name>,
}

impl ModuleDescription<Box<dyn Module>> {
    fn parse(table: &mut NameTable, s: &str) -> Result<(Name, Self)> {
        let (kind_and_name, destinations) = s.split_once(" -> ")
            .ok_or("' -> ' expected")?;
        let destinations: Vec<_> =
            destinations.split(", ").map(|name| table.get(name)).collect();
        let (module, name): (Box<dyn Module>, _) =
            if kind_and_name == "broadcaster" {
                (Box::new(Broadcaster {}), table.get("broadcaster"))
            }
            else {
                let mut chars = kind_and_name.chars();
                let kind: Box<dyn Module> = match chars.next() {
                    Some('%') => Box::new(FlipFlop {}),
                    Some('&') => Box::new(Conjunction::new()),
                    _ => Err(format!("Invalid name {kind_and_name}"))?
                };
                (kind, table.get(chars.as_str()))
            };
        Ok((name, Self { module, destinations }))
    }
}

type Configuration = NameMap<ModuleDescription<Box<dyn Module>>>;

type RunningConfiguration = NameMap<ModuleDescription<Box<dyn RunningModule>>>;

fn run(configuration: &Configuration) -> RunningConfiguration {
    configuration.iter().map(|(&name, description)| {
      let module = description.module.run();
      let destinations = description.destinations.clone();
      (name, ModuleDescription { module, destinations })
    }).collect()
}

struct Message {
    input: Name,
    destination: Name,
    pulse: Pulse,
}

trait Observer {
    fn observe(&mut self, message: &Message);
}

struct State<O: Observer> {
    message_queue: std::collections::VecDeque<Message>,
    observer: O,
}

impl<O: Observer> State<O> {
    fn new(observer: O) -> Self {
        Self {
            message_queue: std::collections::VecDeque::new(),
            observer
        }
    }

    fn send_single(&mut self, message: Message) {
        self.observer.observe(&message);
        self.message_queue.push_back(message)
    }

    fn send(&mut self, input: Name, destinations: &Vec<Name>, pulse: Pulse) {
        for &destination in destinations {
            self.send_single(Message { input, destination, pulse });
        }
    }

    fn push_button(&mut self, table: &mut NameTable) {
        let button = table.get("button");
        let broadcaster = table.get("broadcaster");
        self.send_single(
            Message { input: button, destination: broadcaster, pulse: Pulse::Low }
        );
    }

    fn handle_messages(&mut self, configuration: &mut RunningConfiguration) {
        while let Some(message) = self.message_queue.pop_front() {
            if let Some(description) = configuration.get_mut(&message.destination) {
                description.module.receive(&message, &mut |pulse| {
                    self.send(message.destination, &description.destinations, pulse)
                })
            }
        }
    }
}

struct Counter {
    low: u64,
    high: u64,
}

impl Observer for Counter {
    fn observe(&mut self, message: &Message) {
        match message.pulse {
            Pulse::Low => self.low += 1,
            Pulse::High => self.high += 1,
        }
    }
}

fn part1(table: &mut NameTable, configuration: &Configuration) -> u64 {
    let mut running_configuration = run(&configuration);
    let mut state = State::new(Counter { low: 0, high: 0 });
    for _i in 0 .. 1000 {
        state.push_button(table);
        state.handle_messages(&mut running_configuration);
    }
    state.observer.low * state.observer.high
}

struct FindFirstHighPulses<'a> {
    counter: u64,
    inputs: &'a NameSet,
    indices: NameMap<u64>,
}

impl<'a> Observer for FindFirstHighPulses<'a> {
    fn observe(&mut self, message: &Message) {
        if message.pulse == Pulse::High &&
            self.inputs.contains(&message.input) &&
            !self.indices.contains_key(&message.input)
        {
            self.indices.insert(message.input, self.counter);
        }
    }
}

fn part2(
    table: &mut NameTable, configuration: &Configuration, rx_input: Name
) -> Result<u64> {
    let inputs = configuration.get(&rx_input)
        .ok_or("rx input unconfigured")?
        .module.get_inputs().ok_or("rx input is not conjunction")?;
    let mut running_configuration = run(&configuration);
    let mut state = State::new(FindFirstHighPulses {
            counter: 0, inputs: inputs, indices: NameMap::new()
        });
    while !state.observer.inputs.iter().all(|input|
        state.observer.indices.contains_key(input)
    ) {
        state.push_button(table);
        state.observer.counter += 1;
        state.handle_messages(&mut running_configuration);
    }
    Ok(state.observer.indices.values().cloned().fold(1, lcm))
}

fn main() -> Result<()> {
    let mut table = NameTable::new();
    let mut configuration: Configuration = std::io::stdin().lines()
        .map(|line| -> Result<_> {
            ModuleDescription::parse(&mut table, &line?)
        }).collect::<Result<_>>()?;
    let destinations: Vec<(Name, Vec<Name>)> =
        configuration.iter().map(|(&input, description)|
          (input, description.destinations.clone())).collect();
    let rx = table.get("rx");
    let mut rx_input = None;
    for (input, destinations) in destinations {
        for destination in destinations {
            if let Some(description) = configuration.get_mut(&destination) {
                description.module.add_input(input)
            }
            if destination == rx {
                if rx_input.is_some() {
                    Err("Multiple inputs for rx")?
                }
                rx_input = Some(input)
            }
        }
    }
    let result_part1 = part1(&mut table, &configuration);
    println!("Part 1: {result_part1}");
    let rx_input = rx_input.ok_or("no input for rx")?;
    let result_part2 = part2(&mut table, &configuration, rx_input)?;
    println!("Part 2: {result_part2}");
    Ok(())
}
